### PR TITLE
Add error handler for authed service

### LINKED ISSUE #78: [tsec-http4s] Allow customizing what happens on auth failure
Currently an auth failure results in a 401 response. It can sometimes be useful to instead redirect to some form of login page. This is not possible out of the box, but can be achieved by mapping over the result.

A customization option to allow for changing the failure case would be nice to have.

Maybe something like
```scala
// in SecuredRequestHandler
def liftService(service: TSecAuthService[F, User, Auth], fallback: => Response[F]): HttpService[F] =
  defaultMiddleware(service).handleError(_ => fallback))
```

### LINKED ISSUE #93: Rolling window aid default constructions for apply in authService are _backwards_
Fixing this in a quick PR:

https://github.com/jmcardon/tsec/blob/master/tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala#L59

`afterBlock` handles rolling window updating, which is a no-op for non-rolling window constructions.

however, it's "default" using afterblock, and rolling window not doing it. this is quite literally backwards.

### PR DESCRIPTION
**Pull request by jmcardon**
Customize auth failure and fix rolling window bug

Closes #93 
Closes #78

### CODE REVIEW

File: tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala [hunk 1]
```diff
@@ -10,46 +10,51 @@ sealed abstract class SecuredRequestHandler[F[_], Identity, User, Auth](
     val authenticator: AuthenticatorService[F, Identity, User, Auth]
 )(implicit F: MonadError[F, Throwable]) {
 
-  private[tsec] val defaultFailure: Throwable => F[Response[F]] = _ => F.pure(Response[F](Status.Unauthorized))
+  private[tsec] val catchAllFailures: Throwable => F[Response[F]] = _ => F.pure(Response[F](Status.Unauthorized))
```
Comment (hrhino) [inline tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala]: `... = _ => defaultNotAuthorized`?

File: tsec-libsodium/src/test/scala/tsec/libsodium/authentication/OnetimeAuthTest.scala [hunk 2]
```diff
@@ -26,9 +26,9 @@ class OnetimeAuthTest extends SodiumSpec {
       whenever(s1 != s2) {
         val program: IO[Boolean] = for {
           key <- OnetimeAuth.generateKey[IO]
-          message = s1.utf8Bytes
+          message  = s1.utf8Bytes
           message2 = s2.utf8Bytes
-          tag <- OnetimeAuth.generateTag[IO](key, message)
+          tag  <- OnetimeAuth.generateTag[IO](key, message)
```
Comment (hrhino) [inline tsec-libsodium/src/test/scala/tsec/libsodium/authentication/OnetimeAuthTest.scala]: <3 (although you _could_ align the `=` and `-`)

File: tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala (hunk 1 above)
Comment (jmcardon) [reply tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala]: LOL. good point. brain derp

File: tsec-libsodium/src/test/scala/tsec/libsodium/authentication/OnetimeAuthTest.scala (hunk 2 above)
Comment (jmcardon) [reply tsec-libsodium/src/test/scala/tsec/libsodium/authentication/OnetimeAuthTest.scala]: (I just ran scalafmt, sorry miss jackson `oooo` I am four eeeels)

**Comment by jmcardon** [comment]
@hrhino im willing to provide it as a `Request => F[Response[F]]`, but then at this point, I'm going to compromise and make the catch all return 401.

We're going to run into a really funny situation with the catchall, where we will have to either allow the user to define the generalized handler themselves, or pick a default one.

For now, let's pick the default ourselves.

File: tsec-http4s/src/test/scala/tsec/authentication/RequestAuthenticatorSpec.scala [hunk 3]
```diff
@@ -189,7 +189,7 @@ class RequestAuthenticatorSpec extends AuthenticatorSpec {
 
       response.getOrElse(Response.notFound).map(_.status).unsafeRunSync() mustBe Status.BadGateway
     }
-
+jw
```
Comment (hrhino) [inline tsec-http4s/src/test/scala/tsec/authentication/RequestAuthenticatorSpec.scala]: Por que?

**Comment by jmcardon** [comment]
Alright @rsoeldner @hrhino, after much grokking, here's where I land on this:

- @hrhino's `Request[F] => F[Response[F]]` is definitely what we want. It's sensible to give the extra data of the incoming request to make decisions about what to do about it, in the case of auth failure.
- The catch all portion will simply return `Unauthorized`. In particular, I'll consider adding the catch all as an optional parameter to the method. In particular, catch all _cannot_ be `F[Response[F]]`... this can fail. Catch all is: "well you clearly cannot be trusted to handle your own errors, so we'll be doing it for you".

Let me know where you guys land on this behavior.

File: tsec-http4s/src/test/scala/tsec/authentication/RequestAuthenticatorSpec.scala (hunk 3 above)
Comment (jmcardon) [reply tsec-http4s/src/test/scala/tsec/authentication/RequestAuthenticatorSpec.scala]: laptop fat finger. shush it never happened

File: tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala [hunk 4]
```diff
@@ -10,36 +10,50 @@ sealed abstract class SecuredRequestHandler[F[_], Identity, User, Auth](
...
+  def liftAuthorizedService(
       authorization: Authorization[F, User, Auth],
-      service: TSecAuthService[Auth, User, F]
+      service: TSecAuthService[User, Auth, F],
+      onNotAuthorized: Request[F] => F[Response[F]] = defaultNotAuthorized
   ): HttpService[F] =
-    authorizedMiddleware(authorization)(service)
-      .handleError(_ => Response[F](Status.Unauthorized))
+    authorizedMiddleware(authorization, onNotAuthorized)(service)
+      .handleErrorWith(e => Kleisli.lift(OptionT.pure(cachedUnauthorized)))
```
Comment (hrhino) [inline tsec-http4s/src/main/scala/tsec/authentication/SecuredRequestHandler.scala]: this way LGTM.

**Pull request merged**